tests/object.test.ts

```diff
@@ -0,0 +1,80 @@
+import { Expect, Equal } from '../src/types/helpers';
+import { P, match } from '../src';
+
+describe('Object', () => {
+  it('should match exact object', () => {
+    const fn = () => 'hello';
+
+    const res = match({ str: fn() })
+      .with({ str: 'world' }, (obj) => {
+        type t = Expect<Equal<typeof obj, { str: 'world' }>>;
+        return obj.str;
+      })
+      .with(P.object, (obj) => {
```

Comment:
Could you add test covering how P.object behaves with more inputs:
- Functions
- Primitive values
- Null

It should catch all values that are assignable to the `object` type, and type narrowing and exhaustive should both work

Reply:
I've added test. If you need more, please comment

Reply:
Actually, functions and arrays are part of the `object` type in TypeScript: [Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAkgzgeQEYCsIGNgB4AqA+KAXihyggA9gIA7AEzigHtUNgoB+KYAJwFdoAXFABmAQwA2cCAChZoSFHIBGIrEQtMWOCAC2SRuLzSA9MajmAeuznho5AEyr4yNJoAUAOi+iho6iABKIgI-ECNTcygrWWl5OwBmJ3VXbABtAF1ws0trOMUAFiSXVixQjKzI6NjbRQBWIo00pQAaeOb8zJNsqOsgA)

I think `P.object` should catch anything assignable to `object` to stay close to typescript semantics and play nicely with exhaustiveness checking.

I'm going to update your PR shortly